giskard_vision/core/detectors/metadata_scan_detector.py

```diff
@@ -0,0 +1,154 @@
+
+    def __init__(self, surrogate_function: Callable, metric: Callable, type_task: str) -> None:
+        super().__init__()
+        self.surrogate_function = surrogate_function
+        self.metric = metric
+        self.type_task = type_task
+
+    def get_results(self, model: Any, dataset: Any) -> List[ScanResult]:
+        try:
+            from giskard import Dataset, Model, scan
+        except (ImportError, ModuleNotFoundError) as e:
+            raise GiskardImportError(["giskard"]) from e
+
+        meta = dataset.get_meta(0)
+        list_categories = meta.get_categories()
+        list_metadata = list(meta.get_scannable().keys())
+
+        # If the list of metadata is empty, return no issue
+        if not list_metadata:
+            return []
+
+        # Get dataframe from metadata
+        df_for_scan = self.get_df_for_scan(model, dataset, list_metadata)
+
+        df_for_prediction = df_for_scan.copy()
+
+        def prediction_function(df: pd.DataFrame) -> np.ndarray:
+            return pd.merge(df, df_for_prediction, on="index", how="inner")["prediction"].values
+
+        # Create Giskard dataset and model
+        giskard_dataset = Dataset(df=df_for_scan.copy(), target="target", cat_columns=list_categories + ["index"])
+
+        giskard_model = Model(
+            model=prediction_function,
+            model_type=self.type_task,
+            feature_names=list_metadata + ["index"],
+        )
+
+        # Get scan results
+        results = scan(giskard_model, giskard_dataset)
+
+        list_scan_results = []
+
+        # For each slice found, get appropriate scna results with the metric
+        for issue in results.issues:
+            current_data_slice = giskard_dataset.slice(issue.slicing_fn)
+            filenames = list(current_data_slice.df.sort_values(by="metric", ascending=False)["image_path"].values)
+            list_scan_results.append(
+                self.get_scan_result(
+                    metric_value=current_data_slice.df["metric"].mean(),
+                    metric_reference_value=giskard_dataset.df["metric"].mean(),
+                    metric_name=self.metric.__name__,
+                    filename_examples=filenames,
+                    name=issue.slicing_fn.meta.display_name,
+                    size_data=len(current_data_slice.df),
+                )
+            )
+
+        return list_scan_results
+
+    def get_df_for_scan(self, model: Any, dataset: Any, list_metadata: Sequence[str]) -> pd.DataFrame:
+
+        # Create a dataframe containing each metadata and metric, surrogate target, surrogate prediction
+        # image path for display in html, and index
+        df = {name_metadata: [] for name_metadata in list_metadata}
+        df["metric"] = []
+        df["target"] = []
+        df["prediction"] = []
+        df["image_path"] = []
+        df["index"] = []
+
+        for i in range(len(dataset)):
```

Comment:
why are you by passing the batching here?

we should use model.predict(dataset), which should give a full array of preds based on batchsize,

and concatenate all these into a df instead

Reply:
For each prediction I need the metadata, the labels and the image_path (not implemented for batches), and sometimes the prediction fails for a few examples so it seems hard to use the batch.

Reply:
alright, could you just add a comment here explaining this? and put a todo: batch support? thanks